src/vs/platform/terminal/common/capabilities/shellEnvDetectionCapability.ts

```diff
@@ -59,8 +58,33 @@ export class ShellEnvDetectionCapability extends Disposable implements IShellEnv
 		if (!this._pendingEnv) {
 			return;
 		}
-		this._env = this._pendingEnv;
+		this.applyEnvironmentDiff(this._pendingEnv, isTrusted);
 		this._pendingEnv = undefined;
-		this._onDidChangeEnv.fire(this._env);
+	}
+	// Check for environment differs, and was updated.
+	// This way we only fire an event if the environment actually changed.
+
+	// TODO: Handle deleted variable on TS side. It is already one in bash side
+	// unsetEnvironmentSingleVar()
+	applyEnvironmentDiff(env: Map<string, string>, isTrusted: boolean): void {
+		if (!isTrusted) {
+			return;
+		}
+
+		let envDiffers: boolean = false;
+
+		for (const [key, value] of env.entries()) {
+			if (this._env.has(key) && this._env.get(key) === value) {
+				// Do nothing
+			} else if (value !== undefined) {
+				this._env.set(key, value);
+				envDiffers = true;
+			}
+		}
+
+		if (envDiffers) {
+			this._onDidChangeEnv.fire(this._env);
```

Comment:
We should only fire if we are sure the "new" environment is different from the one we have cached. 
The hope is - this would be less burden on extensions consuming this API.